src/miniflask/state.py

```diff
@@ -394,7 +374,156 @@ def __call__(self, state, event):  # pylint: disable=redefined-outer-name
+
+            # we allow one simple alternative: state[x] if x in state else y
+            if_matches = if_else_regex.findall(self.fn_src.split(":")[1]) if self.fn_src else []
+            if len(if_matches) > 1:
+                raise RegisterError(f"Lambda expression with only one if-else-statement of the form `EXPR(state[x]) if x in state else OTHEREXPR` allowed, but found multiple in: {self.fn_src}")
+
+            # we know parse for lambda expressions of the form:
+            # expr1(x,y,...) if x in state and y in state ... else expr2
+            if len(if_matches) == 1:
+                true_expr_src = if_matches[0][0]
+                false_expr_src = if_matches[0][2]
+                state_cond_src = if_matches[0][1]
+                false_expr_dependencies = [m[1] for m in state_regex.findall(false_expr_src)]
+                for bad_keyword in ["or", "not"]:
+                    if f" {bad_keyword} " in state_cond_src:
+                        raise RegisterError(f"Lambda expression allows only if-else-statements of the form `EXPR(state[x]) if x in state and ... else OTHEREXPR` allowed, but found `{bad_keyword}` in condition of: {self.fn_src}")
+                state_cond_vars = [m[1] for cond_src in state_cond_src.split(" and ") for m in state_in_regex.findall(cond_src)]
+
+                # if the condition is also used in the true_expr_src we can ignore it later to check for its alternatives
+                for state_cond_var in state_cond_vars:
+                    true_expr_regex = re.compile(r"state\s*\[\s*([\"'])" + state_cond_var + r"\1\s*\]")
+                    if true_expr_regex.search(true_expr_src):
+                        self.depends_alternatives[state_cond_var] = false_expr_dependencies
+
+    def str(self):
+        return str(self.varid)
 
     def __str__(self):
         return self.str()
+
+    def __repr__(self):
+        content = [self.str()]
+        if self.fn_src is not None:
+            content.append(f"fn=λ {self.fn_src}")
+        if len(self.depends_on) > 0:
+            content.append(f"depends_on={self.depends_on}")
+        return f"variable({', '.join(content)})"
+
+    # ------------------- #
+    # dependency routines #
+    # ------------------- #
+    @staticmethod
+    def calculate_affects_lists(node_dict):
+        for node in node_dict.values():
+            node.affects = []
+        for node in node_dict.values():
+            for depends_on_varid in node.depends_on:
+                node_dict[depends_on_varid].affects.append(node.varid)
+
+    @staticmethod
+    def topological_sort(node_dict):  # noqa: C901
+        nodes = list(node_dict.values())
+        varid2index = {node.varid: i for i, node in enumerate(nodes)}
+        visited = [False for i in range(len(nodes))]
+        sorted_nodes, cycles, unresolved = [], [], []
+
+        # note: rewrite this recursive DFS to while-loop if RecursionError occurs
```

Comment:
Is this something we can (or should) improve now?

Reply:
I've refactored the recursive call to a stack-based call.
However & I have to admit, the recursive version is much nicer to read & maintain. :sweat_smile: 

What do you think?